File: halgorithm/src/Data/Graph/DepthFirst.hs
module Data.Graph.DepthFirst where

import qualified Data.Map as M
import Data.Graph.Helper
import Data.Misc (exists)

search :: M.Map From Dests -> Vertex -> Route
search field start = _search start [] field

_search :: From -> Route -> M.Map From Dests -> Route
_search _ _ field | M.null field = []
_search curr acc field
    | curr == Goal         = reverse (curr:acc)
    | isDeadEnd || isCleared = backtrack acc
    | otherwise              = case lookup of
                                   Just (nexth:nextt) ->
                                       _search nexth _acc _field
                                       where
                                           _acc   = curr : acc
                                           _field = M.update (\x -> Just nextt) curr field
                                   _                  -> backtrack acc
    where
        lookup         = M.lookup curr field
        isDeadEnd      = lookup == Nothing
        isCleared      = exists curr acc
        backtrack acc_ = case acc_ of
                             (prevh:prevt) -> _search prevh prevt field
                             _             -> []

File: salgorithm/src/main/scala/algorithm/search/graph/DepthFirst.scala
package algorithm.search.graph

import scala.annotation.tailrec

object DepthFirst extends Search {

  override def search(field: Field, start: Vertex, goal: Vertex): Route = _search(field, start, goal, List.empty)

  @tailrec
  private def _search(field: Field, curr: Vertex, goal: Vertex, acc: Route): Route =
    curr match {
      case _ if curr == goal => (curr :: acc).reverse
      case _ if field.isEmpty || acc.isEmpty && (isDeadEnd(field, curr) || isCleared(curr, acc)) => NOT_FOUND
      case _ if isDeadEnd(field, curr) || isCleared(curr, acc) => _search(field, acc.head, goal, acc.tail) // backtrack
      case _ => field.get(curr) match {
        case Some(f) if f.isEmpty && acc.isEmpty => NOT_FOUND
        case Some(f) if f.isEmpty => _search(field, acc.head, goal, acc.tail)
        case Some(f) => _search(field.updated(curr, f.tail), f.head, goal, curr :: acc) // backtrack
        case _ if acc.isEmpty => NOT_FOUND
        case _ => _search(field, acc.head, goal, acc.tail)
      }
  }

}
